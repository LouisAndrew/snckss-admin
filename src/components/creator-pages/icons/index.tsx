import React from 'react'
import { Icon } from '@iconify/react'
import backIcon from '@iconify/icons-mdi/arrow-back'
import addIcon from '@iconify/icons-mdi/add'

import './styles.scss'

interface Props {
        displayAdd: boolean
        goBack: () => void
        add: () => void
}

const Icons: React.FC<Props> = ({ displayAdd, goBack, add }) => {
        return (
                <>
                        <div className="back icon" onClick={goBack}>
                                <Icon icon={backIcon} />
                        </div>
                        {displayAdd && (
                                <div className="add icon" onClick={add}>
                                        <Icon icon={addIcon} />
                                </div>
                        )}
                </>
        )
}

export default Icons
