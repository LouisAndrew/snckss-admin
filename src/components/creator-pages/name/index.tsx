import React from 'react'
import { Input, Label, FormGroup } from 'reactstrap'

interface NameProps {
        name: string
        headerText: string
        placeholderText: string
        handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void
}

const Name: React.FC<NameProps> = ({
        name,
        headerText,
        placeholderText,
        handleChange,
}) => (
        <FormGroup>
                <Label for="name">{headerText}</Label>
                <Input
                        type="text"
                        id="name"
                        onChange={handleChange}
                        placeholder={placeholderText}
                        value={name}
                />
        </FormGroup>
)

export default Name
