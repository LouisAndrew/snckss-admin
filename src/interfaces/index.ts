// import { Category } from './category'
// import { Brand } from './brand'
// import { Order } from './order'
// import { Product } from './product'
// import { Admin } from './admin'

// export { Category, Brand, Order, Product, Admin }
export * from './category'
export * from './brand'
export * from './order'
export * from './product'
export * from './admin'
