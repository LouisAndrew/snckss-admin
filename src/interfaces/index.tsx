import { Category } from './category'
import { Brand } from './brand'
import { Order } from './order'
import { Product } from './product'
import { Admin } from './admin'

export type { Category, Brand, Order, Product, Admin }
