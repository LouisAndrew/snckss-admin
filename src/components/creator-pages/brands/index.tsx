import React, { useState, useEffect } from 'react'
import { useFirestore } from 'reactfire'

import { Brand } from '../../../interfaces/brand'
import List, { ListItem } from '../../list'
import Icons from '../icons'
import './styles.scss'
import BrandEditor from './editor'
import { Category } from '../../../interfaces/category'
import { Product } from '../../../interfaces/product'

const initialBrand: Brand = {
        name: '',
        products: [],
        category: {
                name: '',
                brands: [],
        },
}

interface Props {
        allCategories: Category[]
        allBrands: Brand[]
        allProducts: Product[]
}

/**
 * Almost the same structure as Categoties. Do I need to combine them just as one reusable component?
 * but if yes -> It's going to be a lot of if statements inside on component....
 */
const Brands: React.FC<Props> = ({ allCategories, allBrands, allProducts }) => {
        const [isEditing, setIsEditing] = useState(true)
        const [brd, setBrd] = useState<Brand[]>([])
        const [provideBrand, setProvideBrand] = useState(false)
        const [toProvide, setToProvide] = useState<Brand>(initialBrand)

        const brands = useFirestore().collection('brand')

        // fetch all categories from firestore.
        useEffect(() => {
                setBrd(allBrands)
        }, [])

        const handleRemove = () => {}

        const handleClick = (key: ListItem['key']) => {
                if (!isEditing) {
                        const selectedBrand: Brand = brd.filter(
                                (brand) => brand.name === key
                        )[0]
                        if (selectedBrand) {
                                setIsEditing(true)
                                setProvideBrand(true)
                                setToProvide(selectedBrand)
                        }
                }
        }

        const goBack = () => {
                if (isEditing) {
                        setIsEditing(false)
                        setProvideBrand(false)
                        setToProvide(initialBrand)
                }
        }

        const add = () => {
                if (provideBrand) {
                        setProvideBrand(false)
                }
                setIsEditing(true)
        }

        return (
                <div className="brands">
                        <Icons
                                goBack={goBack}
                                add={add}
                                displayAdd={!isEditing}
                        />
                        <h1>Brands</h1>
                        {isEditing ? (
                                <BrandEditor
                                        brand={toProvide}
                                        providedBrand={provideBrand}
                                        goBack={goBack}
                                        allCategories={allCategories}
                                        allProducts={allProducts}
                                />
                        ) : (
                                <List
                                        headerText=""
                                        items={brd.map((brand) => {
                                                const { name } = brand
                                                const list: ListItem = {
                                                        text: name,
                                                        key: name,
                                                }
                                                return list
                                        })}
                                        handleRemove={handleRemove}
                                        handleClick={handleClick}
                                />
                        )}
                </div>
        )
}

export default Brands
