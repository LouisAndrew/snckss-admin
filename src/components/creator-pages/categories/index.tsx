import React, { useState, useEffect } from 'react'
import { useFirestore } from 'reactfire'

import './styles.scss'
import List, { ListItem } from '../../list'
import { Category } from 'ts/interfaces/category'
import Icons from '../icons'
import CategoryEditor from './editor'
import { Brand } from 'ts/interfaces/brand'
import useSureQuestion from 'hooks/useSureQuestion'
import { Product } from 'ts/interfaces/product'

const initialCategories: Category[] = []
export const initialCategory: Category = {
        name: '',
        // brands: [],
        products: [],
}

interface Props {
        allCategories: Category[]
        allBrands: Brand[]
        allProducts: Product[]
        doRerender: () => void
}

const Categories: React.FC<Props> = ({
        allCategories,
        allBrands,
        allProducts,
        doRerender,
}) => {
        const [isEditing, setIsEditing] = useState(false)
        const [ctg, setCtg] = useState(initialCategories)
        const [provideCategory, setProvideCategory] = useState(false)
        const [toProvide, setToProvide] = useState(initialCategory)

        const categories = useFirestore().collection('categories')
        const products = useFirestore().collection('product')
        const question = useSureQuestion()

        useEffect(() => {
                setCtg(allCategories)
        }, [])

        const createNameDuplicate = (name: string): string => {
                let i = 1
                const names: string[] = ctg.map((category) => category.name)
                while (names.includes(`${name}-${i}`)) {
                        i++
                }

                return `${name}-${i}`
        }

        const handleRemove = (key: any) => {
                const isKeyExists: boolean = allCategories.some(
                        (ctg) => ctg.name === key
                )
                if (isKeyExists) {
                        const categoryRef = categories.doc(key)
                        const toDelete = allCategories.filter(
                                (category) => category.name === key
                        )[0]

                        const deleteCategoryFromProducts = (
                                productName: string
                        ) => {
                                products.doc(productName).update({
                                        category: '',
                                })
                        }

                        const deleteCategory = () => {
                                toDelete.products.forEach((product) => {
                                        const productName = (product as unknown) as string
                                        if (productName) {
                                                deleteCategoryFromProducts(
                                                        productName
                                                )
                                        }
                                })

                                categoryRef.delete().then(() => {
                                        doRerender()
                                })
                        }

                        question.apply(
                                `Are you sure you want to delete ${key}`,
                                'categories-creator',
                                deleteCategory,
                                () => {}
                        )
                }
        }

        const handleClick = (key: ListItem['key']) => {
                if (!isEditing) {
                        const selectedCategory: Category = ctg.filter(
                                (category) => category.name === key
                        )[0]
                        if (selectedCategory) {
                                setIsEditing(true)
                                setProvideCategory(true)
                                setToProvide(selectedCategory)
                        }
                }
        }

        const handleEdit = (key: ListItem['key']) => {
                if (!isEditing) {
                        const selectedCategory: Category = ctg.filter(
                                (category) => category.name === key
                        )[0]

                        if (selectedCategory) {
                                setIsEditing(true)
                                setProvideCategory(true)
                                setToProvide({
                                        ...selectedCategory,
                                        name: createNameDuplicate(
                                                selectedCategory.name
                                        ),
                                })
                        }
                }
        }

        const goBack = () => {
                if (isEditing) {
                        setIsEditing(false)
                        setProvideCategory(false)
                        setToProvide(initialCategory)
                        doRerender()
                }
        }

        const add = () => {
                if (provideCategory) {
                        setProvideCategory(false)
                }
                setIsEditing(true)
        }

        return (
                <div className="categories-creator creator">
                        <Icons
                                goBack={goBack}
                                add={add}
                                displayAdd={!isEditing}
                        />
                        <h1>Categories</h1>
                        {isEditing ? (
                                <CategoryEditor
                                        goBack={goBack}
                                        category={toProvide}
                                        providedCategory={provideCategory}
                                        // allCategories={allCategories}
                                        allBrands={allBrands}
                                        allProducts={allProducts}
                                />
                        ) : (
                                <List
                                        headerText=""
                                        items={ctg.map((category) => {
                                                const { name } = category
                                                const list: ListItem = {
                                                        text: name,
                                                        key: name,
                                                }
                                                return list
                                        })}
                                        handleRemove={handleRemove}
                                        handleClick={handleClick}
                                        handleEdit={handleEdit}
                                />
                        )}
                </div>
        )
}

export default Categories
