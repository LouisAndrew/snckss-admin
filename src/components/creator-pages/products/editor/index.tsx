import React, { useEffect, useState, useReducer } from 'react'

import { Product } from 'ts/interfaces/product'
import { Brand } from 'ts/interfaces/brand'
import { reducer, initialState, Actions } from './reducer'
import SuccessPage from 'components/success'
import { Creations } from 'scene/main/creator'
import Name from 'components/creator-pages/name'
import Desc from './desc'
import ImgUploader from './img-upload'

interface Props {
        providedProduct: boolean
        product: Product
        allBrands: Brand[]
        allProducts: Product[]
        goBack: () => void
}

const ProductEditor: React.FC<Props> = ({
        providedProduct,
        product,
        allBrands,
        allProducts,
        goBack,
}) => {
        const [state, dispatch] = useReducer(reducer, initialState)
        const { name, desc, imgs, success } = state

        const handleChangeName = (
                event: React.ChangeEvent<HTMLInputElement>
        ) => {
                dispatch({
                        type: Actions.SET_NAME,
                        payload: {
                                name: event.target.value,
                        },
                })
        }

        const handleChangeDesc = (
                event: React.ChangeEvent<HTMLInputElement>
        ) => {
                dispatch({
                        type: Actions.SET_DESC,
                        payload: {
                                desc: event.target.value,
                        },
                })
        }

        const handleChangeImgs = (imgUrls: string[]) => {
                dispatch({
                        type: Actions.SET_IMGS,
                        payload: {
                                imgs: imgUrls,
                        },
                })
        }

        const imgUrls: string[] = [
                'https://res.cloudinary.com/dsvdffre0/image/upload/v1594991459/wzv9c2057bbqh6ht8kwi.jpg',
                'https://res.cloudinary.com/dsvdffre0/image/upload/v1594991470/oyjs4mo3xwksv7w7dvsf.png',
        ]

        return !success ? (
                <>
                        <Name
                                name={name}
                                placeholderText="Add product's name here!"
                                headerText="Product's name"
                                handleChange={handleChangeName}
                        />
                        <Desc
                                desc={desc}
                                placeholderText="Add product's description here!"
                                headerText="Product's description"
                                handleChange={handleChangeDesc}
                        />
                        {/* using name as simply a string input here */}
                        <ImgUploader
                                allImgs={imgUrls}
                                handleChange={handleChangeImgs}
                        />
                </>
        ) : (
                <SuccessPage
                        type={Creations.PRODUCT}
                        create={providedProduct}
                />
        )
}

export default ProductEditor
