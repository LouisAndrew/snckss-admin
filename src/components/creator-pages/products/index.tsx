import React, { useState, useEffect } from 'react'

import { Brand } from 'ts/interfaces/brand'
import { Product } from 'ts/interfaces/product'
import { initialBrand } from '../brands'
import List, { ListItem } from 'components/list'
import Icons from '../icons'
import ProductEditor from './editor'

const initialProduct: Product = {
        name: '',
        desc: '',
        PID: '',
        price: 0,
        imgs: [],
        available: false,
        arrivingAt: new Date(),
        timesPurchased: 0,
        multipleVars: false,
        vars: [],
        brand: initialBrand,
}

interface Props {
        allBrands: Brand[]
        allProducts: Product[]
        doRerender: () => void
}

const Products: React.FC<Props> = ({ allBrands, allProducts, doRerender }) => {
        const [isEditing, setIsEditing] = useState(false)
        const [prd, setPrd] = useState<Product[]>([])
        const [provideProduct, setProvideProduct] = useState(false)
        const [toProvide, setToProvide] = useState<Product>(initialProduct)

        useEffect(() => {
                setPrd(allProducts)
        }, [])

        const handleRemove = (key: any) => {}

        const handleClick = (key: ListItem['key']) => {
                if (!isEditing) {
                        const selectedProduct: Product = prd.filter(
                                (product) => product.name === key
                        )[0]
                        if (selectedProduct) {
                                setIsEditing(true)
                                setProvideProduct(true)
                                setToProvide(selectedProduct)
                        }
                }
        }

        const goBack = () => {
                if (isEditing) {
                        setIsEditing(false)
                        setProvideProduct(false)
                        setToProvide(initialProduct)
                        doRerender()
                }
        }

        const add = () => {
                if (provideProduct) {
                        setProvideProduct(false)
                }
                setIsEditing(true)
        }

        return (
                <div className="products-creator creator">
                        <Icons
                                goBack={goBack}
                                add={add}
                                displayAdd={!isEditing}
                        />
                        <h1>Products</h1>
                        {isEditing ? (
                                <ProductEditor />
                        ) : (
                                <List
                                        headerText=""
                                        items={prd.map((product) => {
                                                const { name } = product
                                                const list: ListItem = {
                                                        text: name,
                                                        key: name,
                                                }
                                                return list
                                        })}
                                        handleRemove={handleRemove}
                                        handleClick={handleClick}
                                />
                        )}
                </div>
        )
}

export default Products
