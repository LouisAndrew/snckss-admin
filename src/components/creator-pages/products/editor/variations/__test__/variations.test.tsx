import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import { render, cleanup, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import Variations from '..'
import { Product } from 'ts/interfaces/product'
import { initialProduct } from 'components/creator-pages/products'

// not really much to test here, basically just a select component with some additional handling function
describe('Variation adder on product editor', () => {
        const product1: Product = { ...initialProduct, name: 'mockProduct' }
        let valueContainer: Product[] = []

        const mockHeaderText: string = 'mock'
        const mockAllProducts: Product[] = [product1]
        const mockHandleChange = jest.fn((vars: Product[]) => {
                valueContainer = vars
        })

        afterEach(cleanup)
        const el: React.ReactElement = (
                <Variations
                        headerText={mockHeaderText}
                        allProducts={mockAllProducts}
                        handleChange={mockHandleChange}
                />
        )

        it('renders without crashing', () => {
                const div = document.createElement('div')
                ReactDOM.render(el, div)
        })

        it('should react to user changes correctly', () => {
                const { getByLabelText } = render(el)
                const input: HTMLInputElement = getByLabelText(
                        mockHeaderText
                ) as HTMLInputElement

                fireEvent.change(input, { target: { value: product1.name } })
                expect(mockHandleChange).toHaveBeenCalled()
                expect(valueContainer).toContain(product1)
        })

        it('matches snapshot', () => {
                const tree = renderer.create(el).toJSON()
                expect(tree).toMatchSnapshot()
        })
})
