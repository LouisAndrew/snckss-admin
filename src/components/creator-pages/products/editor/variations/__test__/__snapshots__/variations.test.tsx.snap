// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variation adder on product editor matches snapshot 1`] = `
<div
  className="form-group"
>
  <label
    className=""
    htmlFor="select"
  >
    mock
  </label>
  <select
    className="form-control"
    id="select"
    name="select"
    onChange={[Function]}
    value=""
  >
    <option
      value="default"
    >
      Select multiple
    </option>
    <option
      value="mockProduct"
    >
      mockProduct
    </option>
  </select>
</div>
`;
